encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContextDefaultValueTest.java

```diff
@@ -0,0 +1,185 @@
+      FieldRef.Primitive<Integer> iField = ctx.decodeInteger(FieldDescriptor.of("i"));
+      FieldRef.Primitive<Short> sField = ctx.decodeShort(FieldDescriptor.of("s"));
+      FieldRef.Primitive<Long> lField = ctx.decodeLong(FieldDescriptor.of("l"));
+      FieldRef.Primitive<Double> dField = ctx.decodeDouble(FieldDescriptor.of("d"));
+      FieldRef.Primitive<Float> fField = ctx.decodeFloat(FieldDescriptor.of("f"));
+      FieldRef.Primitive<Boolean> bField = ctx.decodeBoolean(FieldDescriptor.of("b"));
+      FieldRef.Primitive<Character> cField = ctx.decodeChar(FieldDescriptor.of("c"));
+      FieldRef.Boxed<Integer> iiField =
+          ctx.decode(FieldDescriptor.of("ii"), TypeToken.of(Integer.class));
+      FieldRef.Boxed<Short> ssField =
+          ctx.decode(FieldDescriptor.of("ss"), TypeToken.of(Short.class));
+      FieldRef.Boxed<Long> llField = ctx.decode(FieldDescriptor.of("ll"), TypeToken.of(Long.class));
+      FieldRef.Boxed<Double> ddField =
+          ctx.decode(FieldDescriptor.of("dd"), TypeToken.of(Double.class));
+      FieldRef.Boxed<Float> ffField =
+          ctx.decode(FieldDescriptor.of("ff"), TypeToken.of(Float.class));
+      FieldRef.Boxed<Boolean> bbField =
+          ctx.decode(FieldDescriptor.of("bb"), TypeToken.of(Boolean.class));
+      FieldRef.Boxed<Character> ccField =
+          ctx.decode(FieldDescriptor.of("cc"), TypeToken.of(Character.class));
+      FieldRef.Boxed<String> strField =
+          ctx.decode(FieldDescriptor.of("str"), TypeToken.of(String.class));
+      FieldRef.Boxed<Object> objField =
+          ctx.decode(FieldDescriptor.of("obj"), TypeToken.of(Object.class));
+
+      return (creationCtx ->
+          new Foo(
+              creationCtx.getInteger(iField),
+              creationCtx.getShort(sField),
+              creationCtx.getLong(lField),
+              creationCtx.getDouble(dField),
+              creationCtx.getFloat(fField),
+              creationCtx.getBoolean(bField),
+              creationCtx.getChar(cField),
+              creationCtx.get(iiField),
+              creationCtx.get(ssField),
+              creationCtx.get(llField),
+              creationCtx.get(ddField),
+              creationCtx.get(ffField),
+              creationCtx.get(bbField),
+              creationCtx.get(ccField),
+              creationCtx.get(strField),
+              creationCtx.get(objField)));
+    }
+  }
+
+  @Test
+  public void defaultNumericValues_areDecodeCorrectly() throws IOException {
+    Map<Class<?>, ObjectDecoder<?>> objectDecoders = new HashMap<>();
+    objectDecoders.put(Foo.class, new FooObjectDecoder());
+    JsonDataDecoderBuilderContext jsonDataDecoderBuilderContext =
+        new JsonDataDecoderBuilderContext(objectDecoders);
+
+    String json =
+        "{\"i\":null, \"s\":null, \"l\":null, \"d\":null, \"f\":null, \"b\":null, \"c\":null, \"ii\":null, \"ss\":null, \"ll\":null, \"dd\":null, \"ff\":null, \"bb\":null, \"cc\":null, \"str\":null, \"obj\":null}";
```

Comment:
Will defaults work for missing json values, i.e. if the above json did not contain "i" at all, would foo.i equal 0?

Reply:
It won't, but we can support it by adding `getOrDefault` for each `get` in the `CreationContextImpl`.
My concern is that: if it missed the whole entry, will it still be considered as valid?

Reply:
> if it missed the whole entry, will it still be considered as valid?

I think so, otherwise it will not be an optional field afaict. That said, I think it makes sense for `getInt` to return the default value unless the field is marked as required(which I don't think we have a notion for yet), it also makes `getInt` consistent with `get` which returns `null` today

Reply:
Great call! 
Also, since `CreationContext` is the place to have value-related logic, and `DataDecoder` is the place to have decoding logic. 
I will move the default value logic into `CreationContext`, and `DataDecoder` will simply link the `fieldRef` with `null` value when the input missed the  entry or encountered `null` value .